light-client/relayer/src/relayer.rs

```diff
@@ -0,0 +1,163 @@
+                            self.next_block_index,
+                            err
+                        );
+                        std::thread::sleep(Self::ERROR_RETRY_FREQUENCY);
+                    } else {
+                        self.next_block_index += 1;
+                    }
+                }
+            }
+        }
+    }
+
+    fn process_block(&mut self, block_data: &BlockData) -> Result<(), WatcherDBError> {
+        // Tracing
+        let tracer = tracer!();
+
+        let mut span =
+            block_span_builder(&tracer, "poll_block", self.next_block_index).start(&tracer);
+
+        span.set_attribute(TELEMETRY_BLOCK_INDEX_KEY.i64(self.next_block_index as i64));
+
+        // First, check if there's anything interesting in this block
+        let relevant_burns = Self::check_for_relevant_burns(&block_data.contents().outputs);
+
+        // There are no burns with relevant memos in this block so this block is
+        // finished processing.
+        if relevant_burns.is_empty() {
+            return Ok(());
+        }
+
+        // Try to get signatures from the watcher
+        // Loop until we get enough signatures
+        tracer.in_span("loop_for_signatures", |_cx| {
+            loop {
+                let signatures = self.check_for_signatures(self.next_block_index)?;
+                if signatures.len() >= self.config.min_signatures {
+                    self.sender.send(
+                        relevant_burns,
+                        block_data.block(),
+                        block_data.contents(),
+                        signatures,
+                    );
+                    return Ok(());
+                } else {
+                    // We didn't get enough signatures, but let's assume that more are coming soon
+                    // TODO: An alternative here is, actually use a light-client verifier object,
+                    // which is a more correct approach.
+                    log::debug!(
+                        self.logger,
+                        "Did not find a quorum yet for block {}: {} signatures < {} required",
+                        self.next_block_index,
+                        signatures.len(),
+                        self.config.min_signatures
+                    );
+                    std::thread::sleep(Self::POLLING_FREQUENCY);
+                }
+            }
+        })
+    }
+
+    /// Function to match TXOs from a block into interesting vector of
+    /// unspent UTXOs.
+    fn check_for_relevant_burns(outputs: &[TxOut]) -> Vec<TxOut> {
+        // Iterate over each output and filter the results using a parallel iterator.
+        let results: Vec<TxOut> = outputs
+            .into_par_iter()
+            .filter_map(|tx_out| {
+                // View key match against the burn address. If it returns ok, then it's a burn.
+                if tx_out.view_key_match(&burn_address_view_private()).is_ok() {
+                    Some(tx_out.clone())
```

Comment:
As a rule of thumb, I think of starting a thread as a cost on the order of like 1-10 microseconds. But Rayon also has a thread pool underlying it, and I think using rayon is supposed to be considerably faster than `std::thread::spawn`.

The cost of scalar mul in Dalek libraries is like 100 microseconds on fancy intel hardware, and like 1 millisecond on phones.

view key match is doing like 2 scalar muls I think.

So I am fairly confident that this would be better in all cases, but I also think there's a better way to think about it.

1. In cases where the block has only 3 Txos, this is going to be very fast regardless which strategy we pursue.
2. In cases where the blocks have many Txos, the par_iter strategy is going to be much faster

An additional consideration, for me, is that it's really easy code-wise to put `into_par_iter` instead of `into_iter` or whatever, without harming readability, and it's really easy to back out of that later if we have to for some reason. Also, most software that is doing continuous view key matching like this, such as mobilecoind and full-service, are already using into_par_iter in this manner.

The main way I could see us improving this is if we start trying to process multiple blocks in parallel, but we probably shouldn't develop that until it becomes clear that we need that.